src/awkward1/operations/structure.py

```diff
@@ -744,20 +740,49 @@ def concatenate(arrays, axis=0, mergebool=True, highlevel=True):
             + awkward1._util.exception_suffix(__file__)
         )
 
-    batch = [contents[0]]
-    for x in contents[1:]:
-        if batch[-1].mergeable(x, mergebool=mergebool):
-            batch.append(x)
-        else:
-            collapsed = batch[0].mergemany(batch[1:])
-            batch = [collapsed.merge_as_union(x)]
+    if axis < 0:
+        raise ValueError(
+            "the 'axis' of concatenate must be non-negative"
+            + awkward1._util.exception_suffix(__file__)
+        )
+    elif axis == 0:
+        batch = [contents[0]]
+        for x in contents[1:]:
+            if batch[-1].mergeable(x, mergebool=mergebool):
+                batch.append(x)
+            else:
+                collapsed = batch[0].mergemany(batch[1:])
+                batch = [collapsed.merge_as_union(x, 0)]
+
+        out = batch[0].mergemany(batch[1:])
+    else:
+        length = len(contents[0])
+        if any(len(lst) != length for lst in contents[1:]):
+            raise ValueError("all arrays must have the same length for concatenate in axis > 0"
+                + awkward1._util.exception_suffix(__file__)
+            )
+
+        def getfunction(inputs, depth):
+            if depth == axis:
+                arr = inputs
+                batch = [arr[0]]
+                for x in arr[1:]:
+                    collapsed = batch[0].mergemany_as_union(batch[1:], 1)
+                    batch = [collapsed.merge_as_union(x, 1)]
+                out = batch[0].mergemany_as_union(batch[1:], 1)
+                return lambda: (
+                    out,
+                )
```

Comment:
Just

```python
return lambda: out
```

is fine. This looks like something that an auto-formatter like Black went nuts on.

Reply:
I think, it expects a tuple as an output:
```python
awkward1/operations/structure.py:778: in concatenate
    out = awkward1._util.broadcast_and_apply(arrays, getfunction, behavior)[0]
awkward1/_util.py:975: in broadcast_and_apply
    out = apply(broadcast_pack(inputs, isscalar), 0)
...
>               assert isinstance(outcontent, tuple)
E               AssertionError

awkward1/_util.py:746: AssertionError

```

Reply:
You're right! It should be

```python
return lambda: (out,)
```

Python's tuple syntax can be subtle (because trailing commas are ignored in argument lists). But anyway, it doesn't need to be three lines.

Reply:
done